dm/syncer/ddl.go

```diff
@@ -192,6 +195,72 @@ func (s *Syncer) clearOnlineDDL(tctx *tcontext.Context, targetTable *filter.Tabl
 	return nil
 }
 
+// adjustCollation adds collation for create database and check create table.
+func adjustCollation(tctx *tcontext.Context, ddlInfo *ddlInfo, statusVars []byte, charsetAndDefaultCollationMap map[string]string) {
+	switch createStmt := ddlInfo.originStmt.(type) {
+	case *ast.CreateTableStmt:
+		if createStmt.ReferTable != nil {
+			return
+		}
+		var justCharset string
+		for _, tableOption := range createStmt.Options {
+			// already have 'Collation'
+			if tableOption.Tp == ast.TableOptionCollate {
+				return
+			}
+			if tableOption.Tp == ast.TableOptionCharset {
+				justCharset = tableOption.StrValue
+			}
+		}
+		if justCharset == "" {
+			tctx.L().Warn("detect create table risk which use implicit charset and collation", zap.String("originSQL", ddlInfo.originDDL))
+			return
+		}
+		// just has charset, can add collation by charset and default collation map
+		collation, ok := charsetAndDefaultCollationMap[strings.ToLower(justCharset)]
+		if !ok {
+			tctx.L().Error("not found charset default collation.", zap.String("originSQL", ddlInfo.originDDL), zap.String("charset", strings.ToLower(justCharset)))
+			return
+		}
+		tctx.L().Info("detect create table risk which use explicit charset and implicit collation, we will add collation by SHOW CHARACTER SET", zap.String("originSQL", ddlInfo.originDDL), zap.String("collation", collation))
+		createStmt.Options = append(createStmt.Options, &ast.TableOption{Tp: ast.TableOptionCollate, StrValue: collation})
+
+	case *ast.CreateDatabaseStmt:
+		var justCharset, collation string
+		var err error
+		var ok bool
+		for _, createOption := range createStmt.Options {
+			// already have 'Collation'
+			if createOption.Tp == ast.DatabaseOptionCollate {
+				return
+			}
+			if createOption.Tp == ast.DatabaseOptionCharset {
+				justCharset = createOption.Value
+			}
+		}
+
+		// just has charset, can add collation by charset and default collation map
+		if justCharset != "" {
+			collation, ok = charsetAndDefaultCollationMap[strings.ToLower(justCharset)]
+			if !ok {
+				tctx.L().Error("not found charset default collation.", zap.String("originSQL", ddlInfo.originDDL), zap.String("charset", strings.ToLower(justCharset)))
+				return
+			}
+			tctx.L().Info("detect create database risk which use explicit charset and implicit collation, we will add collation by SHOW CHARACTER SET", zap.String("originSQL", ddlInfo.originDDL), zap.String("collation", collation))
+		} else {
+			// has no charset and collation
+			// add collation by server collation from binlog statusVars
+			collation, err = event.GetServerCollationByStatusVars(statusVars)
+			if err != nil {
+				tctx.L().Warn("found error when get collation_server from binlog status_vars, auto use default collation of mysql", zap.Error(err))
```

Comment:
This shouldn't happen in a normal binlog event, so if this actually happened, I think we can just skip here.

Reply:
ok